### PR TITLE
DEV-2987: Warn when a source change is covered only by a visual spec

### PR DESCRIPTION
**Pull request by danielzytohoc**
### Context

The PR adds a fifth advisory detector to the test-presence gate: **`visual-only-coverage`**, which warns when a source change's only coverage is a spec under `visual-tests/`.

**Why.** The presence gate accepts any `*.spec.ts` as coverage, and a visual spec is one. That is correct as a floor — a screenshot is a test — but it made a screenshot the cheapest way to satisfy a blocking gate, and a screenshot proves pixels, not behavior: it has no assertion of its own, it cannot fail on a wrong value, and it passes whenever the picture merely looks the same. The visual-suite overhaul report found four source-touching commits since 2025 that shipped a visual spec as their only test. This detector names that shape at the moment it happens, on the PR, as a warning — the gate's verdict is untouched.

**What it does.** `visualOnlyCoverage(changes)` in `.github/scripts/lib/presence-warnings.mjs` reads the `--name-status` change list the CLI already passes (never the diff, so `isAdvisoryPath()` is untouched and its existing pin keeps holding). It fires when at least one change is production source, at least one change is gate coverage, and every coverage change is a visual spec — on `A`, `M` and `R`, never `D`. Silent when a unit test, a `tests/e2e` spec or a wrapper spec accompanies the change; silent on a visual-only change set with no source (the shape of the upcoming codemod PRs, pinned as the example). The message says what a screenshot does and does not prove and points at `visual-tests/AGENTS.md → Decision rule`, the heading the decision-rule guardrail (G1) creates in #13581. In CI it reaches the sticky comment and a `::warning title=Test-presence gate (visual-only-coverage)::` annotation — the CLI already annotates every warning type, now pinned end to end.

**How the month-later decision gets made.** The report's recommendation was "warn for a month, then decide with the data". `.ai/LOCAL-ENFORCEMENT.md`'s advisory paragraph now carries the criterion and the `gh` recipe to tally it: count PRs that merged with the annotation still standing on their final head SHA; if any did, narrow `COVERAGE_ANY_STATUS` so a visual spec stops counting on its own; a month of zero means keep counting. The recipe reads PR head SHAs because the `Checks / test presence` job is PR-only — the develop squash commit carries a skipped run with no annotations.

**Pinned:** `presence-warnings.test.mjs` gains the detector's cases (fires; each paired-coverage silence with the fixture proven to be gate coverage, so the silence is not vacuous; visual-only change set; codemod shape; modified visual spec alone still fires; deleted spec ignored), and a doc pin that every prose enumeration of the detectors — lib header, CLI header, `checks.yml` comment, `LOCAL-ENFORCEMENT.md`, `TESTING.md` — names the fifth one, each anchored to its own enumeration rather than matched anywhere in the file. `presence-gate-cli.test.mjs` pins the annotation title.

**Review round (second commit).** Three independent reviews before this was opened. Two of them found the same major independently: the doc pin that keeps the five detector enumerations in sync was a whole-file regex, so on three of the five surfaces the enumeration could be deleted while another mention of the string kept the test green. Each surface is now matched inside its own region — the lib header list, the CLI header sentence, the `checks.yml` comment, the `LOCAL-ENFORCEMENT.md` list, the `TESTING.md` sentence — and each anchored match was shown red with that enumeration alone removed. The third major was a false sentence: the tally recipe's rationale said develop's squash commit "carries no check runs"; it carries 49, including a *skipped* `Checks / test presence` run — the job is PR-only, which is the actual reason the recipe reads PR head SHAs. Reworded to that. The verifier replayed every probe on the fixed branch: no regression, 1027 tooling tests green. Remaining nits are judgement calls recorded in the fix commit's body (two enumeration pins reject a legitimately appended sixth detector; the lib-header pin fixes the em-dash glyph).

### Test evidence (required for source changes)

Not a source change (`.github/scripts/**`, `.ai/`, `handsontable/.ai/`, one workflow comment). Presence gate passes as tooling-only.

- `.github/scripts/__tests__/presence-warnings.test.mjs` — detector cases + doc pin, each shown red without the change.
- `.github/scripts/__tests__/presence-gate-cli.test.mjs` — the `::warning` title, end to end.

### Commands run

```
npm run test:tooling
npm run eslint
npm run in visual-tests lint
```

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. DEV-2987

### Affected project(s):

- [ ] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://cla.handsontable.com/sign) — one signature covers both Handsontable and HyperFormula; the `cla/signed` check on this PR confirms it.
- [x] My change requires a change to the documentation.
- [ ] MANUAL QA NEEDED — 

ClickUp task: https://app.clickup.com/t/9015210959/DEV-2987

---

> [!NOTE]
> **Low Risk**
> Warn-only presence-gate advisory and documentation; no changes to blocking gate logic or runtime library code.
> 
> **Overview**
> Adds a fifth **non-blocking** advisory to the test-presence gate: **`visual-only-coverage`**. When production source changes and every file the gate treats as *coverage* is a capture spec under `visual-tests/tests/` (added, modified, or renamed), CI prints a warning and a `Test-presence gate (visual-only-coverage)` check annotation. The blocking verdict is unchanged—a visual `.spec.ts` still satisfies the gate; the nudge is to add behavioral coverage in `tests/e2e/` or unit tests.
> 
> Implementation is `visualOnlyCoverage()` on the existing `git diff --name-status` list (not the unified diff), with `VISUAL_SPEC_RE` scoped to `visual-tests/tests/*.spec.ts` so unit tests and other paths are not misclassified. Docs in `.ai/LOCAL-ENFORCEMENT.md` and `handsontable/.ai/TESTING.md` describe the rule and a one-month **`gh` tally** recipe to decide later whether visual specs should stop counting as sole coverage.
> 
> Tests cover detector edge cases, CLI/`readChanges()` rename parsing, annotation title pinning, and region-sliced doc enumerations via `visual-only-coverage-pins.test.mjs`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c90587e51a6042b7b4b2379d1dcf853f359a7163. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-2797 Visual tests: filters screenshots flake, one different shot per run](https://app.clickup.com/t/9015210959/DEV-2797)
Task linked: [DEV-2986 Visual guardrails: the decision rule and the enforcement map (G1, G7)](https://app.clickup.com/t/9015210959/DEV-2986)
Task linked: [DEV-2987 Visual guardrails: presence-gate warning when only a screenshot covers a source change (G6)](https://app.clickup.com/t/9015210959/DEV-2987)

File: handsontable/.ai/TESTING.md [hunk 1]
```diff
@@ -311,6 +311,8 @@ Machine-enforced by the presence gate (`.github/scripts/test-presence-gate.mjs`)
 
 **New spec vs modify existing:** new public API / plugin / editor → a new spec; a bug fix → add a case to the closest existing spec (its test must fail without the fix).
 
+**The gate warns when a visual spec is the only test change:** a source change whose only test change is a capture spec under `visual-tests/tests/` draws the non-blocking `visual-only-coverage` advisory under the gate's verdict. The criterion that decides whether such a spec keeps counting as coverage, and the recipe that tallies the warning, are in `.ai/LOCAL-ENFORCEMENT.md`.
```
Comment (demtario) [inline handsontable/.ai/TESTING.md]: **medium** — the "only test change" wording is narrower than what the detector does. `visualOnlyCoverage` fires when *every change the gate counts as coverage* is a capture spec, not when the visual spec is the *only* test change: a deleted `tests/e2e` spec or a **new** Jasmine `*.spec.js` (neither is coverage to the gate) can sit right beside the capture spec and the advisory still fires. The lib comment in `collectWarnings()` explicitly rejects this phrasing, and `.ai/LOCAL-ENFORCEMENT.md` states it correctly ("every change the gate counts as coverage"). A reader here would wrongly conclude a co-existing new Jasmine spec suppresses the warning. Reword to the "every counted-coverage change" framing.

File: .github/scripts/lib/presence-warnings.mjs [hunk 2]
```diff
@@ -53,6 +64,25 @@ const TEST_RTL_RE = /rtl|layoutDirection/i;
...
+ * widen. The suffix is deliberate, not a package prefix: a `*.unit.js` for the
+ * visual package's own lib is behavioral coverage, not a capture, and must pair
+ * a source change rather than draw this warning (none exists today, but the G3
+ * and G5 guardrails add lib code there). Deliberately not `docs/tests/` either:
+ * its screenshot spec counts as coverage for a core change too, but a docs spec
+ * beside a core change is not the shape this warning is about, and admitting it
+ * invites the false positive that gets a hook disabled. Exported so the tests
+ * assert the matcher instead of re-deriving it.
+ */
+export const VISUAL_SPEC_RE = /^visual-tests\/tests\/.*\.spec\.ts$/;
```
Comment (demtario) [inline .github/scripts/lib/presence-warnings.mjs:84]: **medium** — `VISUAL_SPEC_RE` is the one load-bearing assumption in this PR with no anti-drift pin. Everything else the month-later tally depends on is pinned precisely because a silent drift would read as a false zero: the annotation title, the check-run name, `headRefOid`, and the PR-only `if:`. This matcher has the same failure mode. Its own comment notes `tests/playwright.config.ts` reserves `tests/visual/` as a future home for visual specs; if the population moves there, the regex silently stops matching, the detector goes permanently quiet, and "a month of zero means keep counting" reads a false zero — no test fails. Add a test that globs the real `visual-tests/**/*.spec.ts` population (all 112 currently under `visual-tests/tests/`) and asserts every file matches `VISUAL_SPEC_RE`. In-idiom with `visual-only-coverage-pins.test.mjs`, which already reads real repo files.

File: .github/scripts/__tests__/presence-gate-cli.test.mjs [hunk 3]
```diff
@@ -174,3 +186,40 @@ test('a page object change is advisory input, not coverage: the blocking verdict
...
 });
+
+// A source line with no RTL logic, so the RTL detector stays quiet and the
+// visual-only-coverage advisory is the only one the scenario can print.
+const PLAIN_SOURCE_LINE = '  scrollTo() { this.render(); }\n';
+const VISUAL_SPEC_BODY = 'test(__filename, async({ tablePage }) => {\n'
+  + '  await tablePage.screenshot({ path: helpers.screenshotPath() });\n'
+  + '});\n';
+
+test('a source change covered only by a new visual spec passes the verdict and raises visual-only-coverage, annotation included', () => {
```
Comment (demtario) [inline .github/scripts/__tests__/presence-gate-cli.test.mjs:243]: **low** — this end-to-end scenario exercises only status `A` (a newly added visual spec). The `M` and `R` shapes reach `visualOnlyCoverage` only through hand-built `{status, path}` fixtures in `presence-warnings.test.mjs`, which never touch `readChanges()`'s real `git diff --name-status` parsing (the `status[0]` truncation and the old/new `\t` split for renames). The parser is correct today, so this is coverage-completeness, not a bug: consider adding two `buildRepo`-style cases — modify an existing base-committed visual spec (`M`), and `git mv` one under `visual-tests/tests/` (`R`) — asserting the CLI still emits the advisory with the right `(M)`/`(R)` suffix and, for the rename, the new path.

**Review by demtario — COMMENTED** [review]
5 inline notes below.

File: .github/scripts/__tests__/presence-warnings.test.mjs [hunk 4]
```diff
@@ -370,6 +372,161 @@ test('Walkontable routing ignores changes outside the engine source, including i
   assert.equal(walkontableRouting([]), null);
 });
 
+// --- visualOnlyCoverage ---
+// The #12086 shape — the one commit in 1244 first-parent commits since
```
Comment (demtario) [inline .github/scripts/__tests__/presence-warnings.test.mjs]: **low** — commit-count mismatch. This comment says `1244` first-parent commits since 2026-03-01; `presence-warnings.mjs:380` and `.ai/LOCAL-ENFORCEMENT.md:155` both say `1251` for the same measurement (base `06b74cfcd`, same date range). One is stale — reconcile so the "measured once" story stays internally consistent.

File: .github/scripts/lib/presence-warnings.mjs [hunk 5]
```diff
@@ -382,6 +469,24 @@ export function collectWarnings({ changes = [], diff = '', prBody } = {}) {
...
+      type: 'visual-only-coverage',
+      // One line, no backticks or asterisks: the CLI's annotation() strips both
+      // before the text becomes a `::warning`, and the pointer has to survive it.
+      // "Every change the gate counts as coverage", not "the only test change":
+      // a deleted tests/e2e spec or a NEW Jasmine spec can sit beside the capture
+      // spec — neither is coverage to the gate, and the detector fires anyway.
+      message: 'Source changed and every change the gate counts as coverage is a visual spec under '
+        + 'visual-tests/tests/. A screenshot proves pixels, not behavior — add the Playwright assertion in '
+        + 'tests/e2e/ (or a unit test) that would fail if the behavior broke, and keep the visual spec for what '
+        + 'only pixels can show. Rule: visual-tests/AGENTS.md → Decision rule.',
```
Comment (demtario) [inline .github/scripts/lib/presence-warnings.mjs:491]: **low** — the warning message points authors to `visual-tests/AGENTS.md → Decision rule`, but no `Decision rule` heading exists in `visual-tests/AGENTS.md` on this branch. The commit discloses that the heading lands with a sibling PR in the series, so this is a merge-order caveat rather than a defect: if this PR merges before that one, the advisory sends authors to a heading that is not there yet. Worth sequencing the merges, or landing a stub heading here.

**Review by demtario — APPROVED** [review]
## Hot-review (demtario) — APPROVE

Rechecked HEAD `c90587e`. Prior Mediums (TESTING.md wording, `VISUAL_SPEC_RE` population pin) and Lows (CLI `M`/`R`, count 1251) look fixed; all five earlier threads resolved. No new Critical/High.

Residual Low only: the advisory still points at `visual-tests/AGENTS.md → Decision rule`, which lands with still-open #13581 — fine to sequence after that merge. LGTM.
